tools/engine_tool/lib/src/typed_json.dart

```diff
@@ -0,0 +1,299 @@
+import 'dart:convert' show JsonEncoder, jsonDecode;
+
+import 'package:meta/meta.dart';
+
+/// A JSON object that contains key-value pairs of other JSON values.
+///
+/// This extension type provides type-safe access to the values in the
+/// underlying JSON object without the need for manual type casting and with
+/// helpful APIs for common operations.
+///
+/// ## Examples
+///
+/// ### Basic Usage
+///
+/// ```dart
+/// const json = JsonObject({
+///  'name': 'John Doe',
+///  'age': 30,
+///  'isEmployed': true,
+///  'pets': ['Fido', 'Whiskers'],
+/// });
+///
+/// final name = json.string('name');
+/// final age = json.integer('age');
+/// final isEmployed = json.boolean('isEmployed');
+/// final pets = json.stringList('pets');
+/// ```
+///
+/// ### Null Safety
+///
+/// In the above example, we assumed that all keys exist in the JSON object. To
+/// gracefully handle cases where a key may not exist, you can use the `OrNull`
+/// variants of the methods:
+///
+/// ```dart
+/// final name = json.stringOrNull('name');
+/// ```
+///
+/// This works nicely with null-aware operators:
+///
+/// ```dart
+/// final name = json.stringOrNull('name') ?? 'Unknown';
+/// ```
+///
+/// ### Bulk Operations
+///
+/// By default, a [JsonReadException] is thrown if a key is not found or if the
+/// value cannot be cast to the expected type, which can be handled using a
+/// `try-catch` block:
+///
+/// ```dart
+/// try {
+///   final name = json.string('name');
+/// } on FormatException catch (e) {
+///   print('Error: $e');
+/// }
+/// ```
+///
+/// However, it is cumbersome to wrap every access in a `try-catch` block, and
+/// sometimes you will want to collect all errors at once. For this, you can use
+/// [map] to read multiple values at once and convert them to a custom object:
+///
+/// ```dart
+/// try {
```

Comment:
This seems like a good pattern to follow, and I think it's necessary to use it when parsing JSON with this API in the situation that the errors that are accumulated are expected to be user-actionable. (For example, if we used this API for parsing the build config json under ci/builders.)

In this PR, the JSON that this API is parsing is internal. That is, if there's a problem with it, the user wouldn't necessarily be expected to fix it. So if the JSON exceptions aren't caught and make it to the terminal, that seems to be WAI.

But that leaves the awkward situation that the pattern noted here, which should be followed in certain situations, doesn't have a live example in this PR. My suggestion to avoid that awkwardness would be that we just *always* follow this pattern, so that future users of the API follow it as well even when it isn't strictly necessary.

Reply:
> But that leaves the awkward situation that the pattern noted here, which should be followed in certain situations, doesn't have a live example in this PR.

Do you mean that we don't have `try/catch` of `JsonMapException`?

I could imagine doing that, but as you mentioned we don't have any way to do recovery in this CLI - mal-formatted JSON means we'll have to stop, so the best we could do is try-catch and provide a more context-aware error message (for example `'Failed to parse process artifacts: $error``).

Reply:
> the best we could do is try-catch and provide a more context-aware error message (for example `'Failed to parse process artifacts: $error'`).

Yeah, I think if there's a nice place to try that in this PR, then that would be an improvement.

Reply:
Sounds reasonable.

I might add a `onError` callback in that case so nesting doesn't get unnecessarily long.

Reply:
Done, and added as discussed. Thanks!